### PR TITLE
MULE-14133: Change DISCARD and DISCARD_OLDEST behavior so that the http connector does not timeout.

### PR DESCRIPTION
**Pull request by fsgonz**

### CODE REVIEW

File: modules/http/src/main/java/org/mule/module/http/internal/listener/grizzly/GrizzlyRequestDispatcherFilter.java [hunk 1]
```diff
@@ -122,6 +124,19 @@ public void responseReady(HttpResponse httpResponse, ResponseStatusCallback resp
         return ctx.getSuspendAction();
     }
 
+    private void prepareStopAction(final FilterChainContext ctx, final HttpRequestPacket request, HttpStatus status)
```
Comment (dfeist) [inline modules/http/src/main/java/org/mule/module/http/internal/listener/grizzly/GrizzlyRequestDispatcherFilter.java]: This is more sendEmptyResponse() isn't it?

File: core/src/main/java/org/mule/config/pool/MuleDiscardOldestPolicy.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.config.pool;
+
+import java.util.concurrent.RejectedExecutionHandler;
+import java.util.concurrent.ThreadPoolExecutor;
+
+public class MuleDiscardOldestPolicy implements RejectedExecutionHandler
```
Comment (dfeist) [inline core/src/main/java/org/mule/config/pool/MuleDiscardOldestPolicy.java]: Class level javadoc. How does this differ to DiscardOldestPolicy? Can it replace it? If we don't replace it then can we at least extend it?

File: core/src/main/java/org/mule/config/pool/MuleDiscardPolicy.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.config.pool;
+
+import java.util.concurrent.RejectedExecutionHandler;
+import java.util.concurrent.ThreadPoolExecutor;
+
+public class MuleDiscardPolicy implements RejectedExecutionHandler
```
Comment (dfeist) [inline core/src/main/java/org/mule/config/pool/MuleDiscardPolicy.java]: Class level javadoc. How does this differ to DiscardOldestPolicy? Can it replace it? If we don't replace it then can we at least extend it?

File: core/src/main/java/org/mule/config/pool/StoppableRunnable.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.config.pool;
+
+
+public abstract class StoppableRunnable implements Runnable
```
Comment (dfeist) [inline core/src/main/java/org/mule/config/pool/StoppableRunnable.java]: Class-level javadoc.  Also it's more `cancel` than `stop` isn't it? If possible compose Runnable and another interface if one exists.

File: modules/http/src/test/java/org/mule/module/http/functional/listener/HttpListenerWorkerThreadingProfileTestCase.java [hunk 5]
```diff
@@ -76,28 +89,94 @@ public void setup()
...
+
+    private void assertMaxThreadsActive(String flowName, String url, int maxThreadsActive, int expectedHttpStatus) throws Exception
     {
         maxActiveNumberOfRequestExecutedLatch = new CountDownLatch(maxThreadsActive);
 
         sendRequestUntilNoMoreWorkers(flowName, url, maxThreadsActive);
 
         try
         {
+
```
Comment (dfeist) [inline modules/http/src/test/java/org/mule/module/http/functional/listener/HttpListenerWorkerThreadingProfileTestCase.java]: Remove?

File: core/src/main/java/org/mule/config/pool/MuleDiscardOldestPolicy.java (hunk 2 above)
Comment (fsgonz) [reply core/src/main/java/org/mule/config/pool/MuleDiscardOldestPolicy.java]: I've extended DiscardOldestPolicy though in this case no logic can be reused (though it is correct extend that class).

File: core/src/main/java/org/mule/config/pool/MuleDiscardPolicy.java (hunk 3 above)
Comment (fsgonz) [reply core/src/main/java/org/mule/config/pool/MuleDiscardPolicy.java]: I've extended DiscardPolicy though in this case no logic can be reused (though it is correct extend that class).

File: core/src/main/java/org/mule/config/pool/CancellableRunnable.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+package org.mule.config.pool;
+
+/**
+ * <code>CancellableRunnable</code> defines an interface where
+ * some logic can be executed in case the work may be cancelled
+ * by a rejection policy.
+ */
+public abstract class CancellableRunnable implements Runnable
+{
+    public abstract void cancelExecution();
```
Comment (dfeist) [inline core/src/main/java/org/mule/config/pool/CancellableRunnable.java]: minor: -> `cancel()`

File: modules/http/src/main/java/org/mule/module/http/api/HttpConstants.java [hunk 7]
```diff
@@ -78,6 +78,11 @@ public int getDefaultPort()
     {
         public static String GRIZZLY_MEMORY_MANAGER_SYSTEM_PROPERTY = "org.glassfish.grizzly.DEFAULT_MEMORY_MANAGER";
     }
+    
+    public static final class HttpResponseReasonPhrases
+    {
+        public static String DROPPED = "DROPPED";
```
Comment (dfeist) [inline modules/http/src/main/java/org/mule/module/http/api/HttpConstants.java]: Why not add an enum?  `DROPPED(202, "DROPPED")`?

File: modules/http/src/main/java/org/mule/module/http/internal/listener/grizzly/GrizzlyRequestDispatcherFilter.java [hunk 8]
```diff
@@ -122,6 +125,19 @@ public void responseReady(HttpResponse httpResponse, ResponseStatusCallback resp
...
+    {  
+        final HttpResponsePacket.Builder responsePacketBuilder = HttpResponsePacket.builder(request);
+        responsePacketBuilder.status(status.getStatusCode());
+        HttpResponsePacket packet = responsePacketBuilder.build();
+        // this is necessary to avoid chunk-encoding
+        packet.setContentLength(0);
+        packet.setReasonPhrase(reason);
+
+        ctx.write(packet);
+        ctx.notifyDownstream(HttpServerFilter.RESPONSE_COMPLETE_EVENT);
```
Comment (dfeist) [inline modules/http/src/main/java/org/mule/module/http/internal/listener/grizzly/GrizzlyRequestDispatcherFilter.java]: static import?

**Pull request merged**